src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java

```diff
@@ -224,4 +284,67 @@ public String[] getAllQueryIds() {
     public Config getConfig() {
         return config;
     }
+
+
+   /**
+    * Instantiates pattern queries from the given query source by querying a SPARQL endpoint.
+    * A query pattern is a SPARQL 1.1 Query, which can have additional variables in the regex form of
+    * <code>%%var[0-9]+%%</code> in the Basic Graph Pattern.
+    * <p>
+    * Exemplary pattern: </br>
+    * <code>SELECT * WHERE {?s %%var1%% ?o . ?o &lt;http://exa.com&gt; %%var2%%}</code><br/>
+    * This pattern will then be converted to: <br/>
+    * <code>SELECT ?var1 ?var2 {?s ?var1 ?o . ?o &lt;http://exa.com&gt; ?var2}</code><br/>
+    * and will request query solutions from the given sparql endpoint (e.g DBpedia).<br/>
+    * The solutions will then be instantiated into the query pattern.
+    * The result may look like the following:<br/>
+    * <code>SELECT * WHERE {?s &lt;http://prop/1&gt; ?o . ?o &lt;http://exa.com&gt; "123"}</code><br/>
+    * <code>SELECT * WHERE {?s &lt;http://prop/1&gt; ?o . ?o &lt;http://exa.com&gt; "12"}</code><br/>
+    * <code>SELECT * WHERE {?s &lt;http://prop/2&gt; ?o . ?o &lt;http://exa.com&gt; "1234"}</code><br/>
+    */
+    private static List<String> instantiatePatternQueries(QuerySource querySource, Config.Pattern config) throws IOException {
+        final var patternQueries = new FileCachingQueryList(querySource);
+        final Pattern pattern = Pattern.compile("%%var\\d+%%");
+        final var instances = new ArrayList<String>();
+        for (int i = 0; i < patternQueries.size(); i++) {
+            // replace all variables in the query pattern with SPARQL variables
+            // and store the variable names
+            var patternQueryString = patternQueries.getQuery(i);
+            final Matcher matcher = pattern.matcher(patternQueryString);
+            final var variables = new ArrayList<String>();
+            while (matcher.find()) {
+                final var match = matcher.group();
+                final var variable = "?" + match.replaceAll("%%", "");
+                variables.add(variable);
+                patternQueryString = patternQueryString.replaceAll(match, variable);
+            }
+
+            // build SELECT query for finding bindings for the variables
+            final var patternQuery = QueryFactory.create(patternQueryString);
+            final var whereClause = "WHERE " + patternQuery.getQueryPattern();
```

Comment:
This approach does not consider subqueries. 
Please state in the documentation that placeholders in subqueries must be projected from the subquery, otherwise they will not be processed correctly.  

If we really want to support subqueries that needs a bit more thought. We would probably need to extract them from the jena query object and send them separately. Also, we would need to handle deeply nested subqueries somehow. Seems like a deep rabi thole to me that I would like to avoid. 

@aalexandrasilva: Do you have placeholders in subqueries in your workload?

Reply:
No, I don't have placeholders in the subqueries :smile: